### DISCUSSION TITLE
tcmenu compatibility with Lilygo T-embed ESP32-s3 custom control board

### DISCUSSION THREAD

**Discussion by beatlepicker** [davetcc/tcMenu#319] [Q&A] [answered]
Has anyone tried any tcmenu compatibility testing with the above T-embed ESP32-s3 board? It seems well suited as it has an integral rotary encoder with pushbutton, and an LCD using SPI on the ST7789 lcd controller. It kind of looks like it was designed with a tcmenu style UI in mind. I'm thinking of getting one to try with tcmenu and would welcome any words of encouragement etc. regarding compatibility. One thing of note is that this T-embed board already uses most of the existing ESP32-s3 GPIO pins, so there is limited selection for further customisation.

**Comment by davetcc** [comment, marked as answer]
I have tried it with a esp32s3 mini with an external rotary encoder and st7735. It is even in the examples. Should not be difficult to set up for that board.

**Reply by beatlepicker** [reply]
Thanks!

**Comment by AtmosphEng** [comment]
Hi Dave & Co.  I'm currently trying to put Tcmenu on a Lilygo T-Display-S3. This has a 1.9 inch 170x320 LCD connected to a ST7789 controller in 8-bit parallel mode. Do you please have any Tcmenu example code for an ESP32 interfacing in 8-bit parallel mode to the LCD? Or any other pointers where to start?

**Reply by davetcc** [reply]
Sorry no it has not been attempted before to the best of my knowledge.

**Reply by AtmosphEng** [reply]
Thanks

**Reply by davetcc** [reply]
Writing a custom device drawable is not that complex however. That would be what I suggest.

**Reply by davetcc** [reply]
See https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/tcmenu-plugins/customise-menu-input-display-plugin/

**Reply by davetcc** [reply]
Original messages were from my phone while I was out and about, I've taken a look and you may be in luck - I would take a look at https://github.com/Bodmer/TFT_eSPI/issues/557

If you got that working, it would be out of box as you just configure TFT_eSPI and then use the plugin.

**Reply by AtmosphEng** [reply]
Thanks for following-up on this. I'll probably look at this later. One of my problems here was that AliExpress messed up my order, as I actually wanted a more tcmenu compatible ESP32 board with a LCD on the SPI.

**Reply by davetcc** [reply]
But actually, I think it would be compatible on reflection, it may take quite a bit of configuration, but from what I can see it would work in both TFT_eSPI and even _possibly_ Adafruit's library too.

**Reply by AtmosphEng** [reply]
UPDATE: I finally got a hold of a lilygo-t-embed ESP32-S3. The Lilygo t-embed github examples used the LCD TFT_eSPI library, so I used the Designer 3.1.0.  Code Gernerator TFT_eSPI option. The extra step of editing User_setup_Select.h to select Setup210_Lilygo_T_Embed_S3.h is also required, and is suggested by your Code Generator 'Click for documentation' link. Now it works OK with a simple example using the onboard ST7789V controlled 170x320 LCD and onboard rotary encoder. Also works OK with embedControl on Windows 11 Home using the ESP32 Serial1 UART. Thanks.

**Reply by davetcc** [reply]
Thanks for letting me know, good news there.

EmbedControl Should also work with the board's WiFi too should you want to use that too.

**Reply by AtmosphEng** [reply]
FYI: I also tried the same TFT_eSPI LCD method above on the Lilygo ESP32 T_Display-S3. This board has an 8-bit parallel data interface instead of SPI. I enabled the corresponding setup file Setup206_LilyGo_T_Display_s3.h. Unfortunately, although it generated some seemingly random and striped patterns on the LCD, I could not get it to render my simple Tcmenu example.

**Reply by davetcc** [reply]
May be one to bring up in the TFTeSPI forum. Possibly just may need a different configuration

**Reply by AtmosphEng** [reply]
SUCCESS ALSO WITH LILYGO T-DISPLAY-S3 LCD 8-bit parallel data with TFT_eSPI library. I was triple checking my test code for errors before asking the TFT_eSPI forum and I found I was using the wrong pin_config.h file. Now it renders my simple tcmenu example fine. This board is a popular one as Lilygo as they say they have had about 2300 orders - this is about 10x more than their typical order numbers.

**Comment by davetcc** [comment]
Great news.
